src/__tests__/getMarkupFromTree-test.tsx

```diff
@@ -0,0 +1,506 @@
+import ApolloClient from 'apollo-client';
+import { ApolloLink, Observable } from 'apollo-link';
+import { MockedResponse } from 'apollo-link-mock';
+import gql from 'graphql-tag';
+import * as React from 'react';
+import { renderToString } from 'react-dom/server';
+
+import { ApolloProvider } from '../ApolloContext';
+import { TestErrorBoundary } from '../__testutils__/TestErrorBoundary';
+import createClient from '../__testutils__/createClient';
+import { getMarkupFromTree } from '../getMarkupFromTree';
+import { QueryHookOptions, useQuery } from '../useQuery';
+
+const AUTH_QUERY = gql`
+  {
+    isAuthorized
+  }
+`;
+
+interface AuthQueryResult {
+  isAuthorized: boolean;
+}
+
+const USER_QUERY = gql`
+  {
+    currentUser {
+      firstName
+    }
+  }
+`;
+
+interface UserQueryResult {
+  currentUser: { firstName: string };
+}
+
+const MOCKS: MockedResponse[] = [
+  {
+    request: { query: AUTH_QUERY },
+    result: { data: { isAuthorized: true } },
+  },
+  {
+    request: { query: USER_QUERY },
+    result: { data: { currentUser: { firstName: 'James' } } },
+  },
+];
+
+const linkReturningError = new ApolloLink(
+  () =>
+    new Observable(observer => {
+      observer.error(new Error('Simulating network error'));
+    })
+);
+
+function createMockClient(link?: ApolloLink) {
+  return createClient({ link, mocks: MOCKS, addTypename: false });
+}
+
+function useAuthDetails(options?: QueryHookOptions<{}>) {
+  const { data, loading } = useQuery<AuthQueryResult>(AUTH_QUERY, options);
+
+  return Boolean(!loading && data && data.isAuthorized);
+}
+
+function UserDetails(props: QueryHookOptions<{}>) {
+  const { data, loading } = useQuery<UserQueryResult>(USER_QUERY, props);
+
+  return (
+    <>
+      {loading
+        ? 'Loading'
+        : !data
+        ? 'No Data (skipped)'
+        : !data.currentUser
+        ? 'No Current User (failed)'
+        : data.currentUser.firstName}
+    </>
+  );
+}
+
+interface UserWrapperProps extends QueryHookOptions<{}> {
+  readonly client: ApolloClient<object>;
+}
+
+function UserDetailsWrapper({ client, ...props }: UserWrapperProps) {
+  return (
+    <TestErrorBoundary>
+      <ApolloProvider client={client}>
+        <UserDetails {...props} />
+      </ApolloProvider>
+    </TestErrorBoundary>
+  );
+}
+
+it('should run through all of the queries that want SSR with suspense', async () => {
```

Comment:
The tests are actually the same for suspene / non-suspense versions. We could make it more DRY with something like that:

```typescript
[true, false].forEach(suspend => {
  const prefix = suspend ? 'with' : 'without';

  describe(`${prefix} suspense`, () => {
    it('should run through all of the queries that want SSR with suspense', async () => {
      const client = createMockClient();

      await expect(
        getMarkupFromTree({
          renderFunction: renderToString,
          tree: <UserDetailsWrapper client={client} suspend={suspend} />,
        })
      ).resolves.toBe('James');
    });

    // the same for other tests
  });
});
```

Inline snapshots wouldn't work in that case but it shouldnt't be a problem - we were checking simple strings so we could easily replace `toMatchInlineSnapshot` with `toBe`.

Reply:
I used [describe.each](https://jestjs.io/docs/en/api#describeeachtable-name-fn-timeout) instead, and all tests MUST have similar output so `toMatchInlineSnapshot` works just fine.

Reply:
Good idea.